HIP/hip-000.md

```diff
@@ -0,0 +1,116 @@
+
+Ethereum nodes expose a JSON-RPC endpoint that can be queried for each account’s nonce value: [eth_getTransactionCount](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactioncount)
+
+In Hedera there is `JSON-RPC Relay` (as described in [HIP-482](https://hips.hedera.com/hip/hip-482)) that provides the same `eth_getTransactionCount` JSON-RPC endpoint, but currently it always returns a value of 1 for contract accounts. The desired behavior would be to have a nonce value of `1 + n` where `n` is the number of times the contract created other contracts through `CREATE` or `CREATE2` operations.
+
+## Rationale
+
+As described in [HIP-482](https://hips.hedera.com/hip/hip-482), the `JSON-RPC Relay` works with `EthereumTransactions` that it forwards to the consensus nodes. Apart from the `EthereumTransaction` type the consensus nodes can also handle other types of transactions (e.g. `ContractCall` or `ContractCreate`) that are not coming from the `JSON-RPC Relay` but can be used by developers by other means.
+In general, when we talk about account nonces we will need to consider all of those transaction types for both types of accounts (i.e. EOAs and contracts):
+
+- Contract nonce updates for both cases of `EthereumTransactions` and `non-EthereumTransactions` (i.e. `ContractCall` or `ContractCreate`) will need to be tracked and externalized to Mirror Node
+- Tracking nonces for EOAs for `EthereumTransactions` is not changed (currently we don't have a specific protobuf to externalize the nonce but the nonce value is passed as part of the `EthereumTransaction` body, so Mirror Node can read the value from there)
+- Tracking nonces for EOAs for `non-EthereumTransactions` is not supported because there is no use case for it in Hedera
+
+## User stories
+
+1. As a user, when I make a request to `eth_getTransactionCount` endpoint I want to be able to get correct nonce values for contracts.
+  
+## Specification
+
+No changes to Services state will be needed because currently the nonces are currently being computed correctly there. We need to externalize these changes to Mirror Node so that it can keep a correct state representation for contract accounts.
+
+### HAPI Changes
+
+Define new message type:
+
+```protobuf
+message ContractNonceInfo {
+
+    /**
+     * Id of the contract
+     */
+    ContractID contract_id = 1;
+
+    /**
+     * The current value of the contract account's nonce property
+     */
+    int64 nonce = 2;
+}
+```
+
+Update `ContractFunctionResult`:
+
+```protobuf
+message ContractFunctionResult {
+    ...
+
+    /**
+     * A list of updated contract account nonces containing the new nonce value for each contract account
+     */
+    repeated ContractNonceInfo contract_nonces = 14;
```

Comment:
There was a reason why we can't track this on each child's function result and just have a `ContractFunctionResult.nonce`, but I don't remember what it was. Can you clarify why by updating the docs?

Reply:
Had a discussion with @stoqnkpL about that earlier today and the main consideration regarding the `ContractFunctionResult.nonce` approach is that in `ContractFunctionResult` we have the contract that is the target but we don't have the sender (which is the one that should have it's nonce incremented).
Also in the case of revert in some of the child transactions we are not sure if there will be a record produced for them, that's why I think the `contract_nonces` should be populated in the parent record.
I will update docs with this information.

Reply:
Added to `Rejected Ideas` section.